managedservices/managedservices.go

```diff
@@ -0,0 +1,230 @@
+package managedservices
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/giantswarm/apprclient"
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	frameworkresource "github.com/giantswarm/e2e-harness/pkg/framework/resource"
+	"github.com/giantswarm/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Config struct {
+	Namespace string
+
+	ApprClient    *apprclient.Client
+	HelmClient    *helmclient.Client
+	HostFramework *framework.Host
+	Logger        micrologger.Logger
+}
+
+type ManagedServices struct {
+	namespace string
+
+	apprClient    *apprclient.Client
+	helmClient    *helmclient.Client
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+	resource      *frameworkresource.Resource
+}
+
+func New(config Config) (*ManagedServices, error) {
+	var err error
+
+	if config.Namespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Namespace must not be empty", config)
+	}
+
+	if config.ApprClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ApprClient must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.HelmClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HelmClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	var resource *frameworkresource.Resource
+	{
+		c := frameworkresource.ResourceConfig{
+			ApprClient: config.ApprClient,
+			HelmClient: config.HelmClient,
+			Logger:     config.Logger,
+			Namespace:  config.Namespace,
+		}
+
+		resource, err = frameworkresource.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	ms := &ManagedServices{
+		apprClient:    config.ApprClient,
+		helmClient:    config.HelmClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		resource:      resource,
+	}
+
+	return ms, nil
+}
+
+func (ms *ManagedServices) Test(ctx context.Context, chartConfig ChartConfig, chartResources ChartResources) error {
```

Comment:
Why aren't the other arguments besides the context be part of the test initialization? IIRC so far the interface is always the same and the test instance is just configured with everything.

Reply:
I wasn't sure about that and keeping the interface makes sense to me. I'll fix.